lib/logstash/outputs/logstash.rb

```diff
@@ -65,125 +81,223 @@ def initialize(*a)
-      }
+      fail(LogStash::ConfigurationError, "`ssl_keystore_password` is REQUIRED when `ssl_keystore_path` is provided") if @ssl_keystore_password.nil?
     elsif @ssl_keystore_password
       fail(LogStash::ConfigurationError, "`ssl_keystore_password` is not allowed unless `ssl_keystore_path` is configured")
     else
-      return {}
+      # acceptable
     end
   end
 
-  def ssl_trust_options
-    {
-      'ssl_verification_mode' => @ssl_verification_mode,
-    }.tap do |trust_options|
-      if @ssl_certificate_authorities&.any? && @ssl_truststore_path
-        fail(LogStash::ConfigurationError, 'SSL trust can be configured with EITHER `ssl_certificate_authorities` OR `ssl_truststore_*`, but not both')
-      elsif @ssl_certificate_authorities&.any?
-        fail(LogStash::ConfigurationError, 'SSL Certificate Authorities cannot be configured when `ssl_verification_mode => none`') if @ssl_verification_mode == 'none'
+  def validate_ssl_trust_options!
+    if @ssl_certificate_authorities&.any? && @ssl_truststore_path
+      fail(LogStash::ConfigurationError, 'SSL trust can be configured with EITHER `ssl_certificate_authorities` OR `ssl_truststore_*`, but not both')
+    elsif @ssl_certificate_authorities&.any?
+      fail(LogStash::ConfigurationError, 'SSL Certificate Authorities cannot be configured when `ssl_verification_mode => none`') if @ssl_verification_mode == 'none'
+    elsif @ssl_truststore_path
+      fail(LogStash::ConfigurationError, 'SSL Truststore cannot be configured when `ssl_verification_mode => none`') if @ssl_verification_mode == 'none'
+      fail(LogStash::ConfigurationError, '`ssl_truststore_password` is REQUIRED when `ssl_truststore_path` is provided') if @ssl_truststore_password.nil?
+
+    elsif @ssl_truststore_password
+      fail(LogStash::ConfigurationError, '`ssl_truststore_password` not allowed unless `ssl_truststore_path` is configured')
+    end
+  end
+
+  def construct_host_uri
+    scheme = @ssl_enabled ? 'https'.freeze : 'http'.freeze
+    [].tap do |downstream_uris|
+      @hosts.each do | host_port_pair | # Struct(:host, :port)
+        uri = LogStash::Util::SafeURI.new(host_port_pair[:host])
+        uri.port = host_port_pair[:port].nil? ? DEFAULT_PORT : host_port_pair[:port]
+        uri.update(:scheme, scheme)
+        # we only need `SafeURI::String` to directly apply to Manticore client
+        downstream_uris << uri.to_s.freeze
+      end
+    end
+  end
 
-        trust_options['ssl_certificate_authorities'] = @ssl_certificate_authorities.first
-      elsif @ssl_truststore_path
-        fail(LogStash::ConfigurationError, 'SSL Truststore cannot be configured when `ssl_verification_mode => none`') if @ssl_verification_mode == 'none'
+  def send_events(events)
+    # we use array to utilize `peek` feature, note that every thread has its own array here
+    pending = Queue.new
```

Comment:
A lot of the details here are the remnant of "hacking" whole-batch processing into the Http output's existing send queue, and without the un-batched processing we are left with a lot of details that could be significantly simplified:

pseudo-code:

~~~ ruby
  def send_events(events)
    # build and compress the body just once
    body = LogStash::Json.dump(events.map(&:to_hash))
    compressed_body = gzip(body)
    
    begin
      @load_balancer.select do |selected_host_uri|
        # use `Manticore::Client#post(...)` instead of `Object#send("post",...)`
        response = client.post(selected_host_uri, :body => compressed_body, :headers => @headers).call
        # fail _inside_ the `LoadBalancer#select` unless the downstream
        # successfully received the events so that our load balancer can
        # be aware of the failure 
      end
    rescue => exception
      # handle _all_ failures here, using the `retry` keyword to retry unless the pipeline is being shutdown.
      
    end
  end

Reply:
As we discussed, the `retry` keyword in ruby can only be used in a `rescue` clause, so if we want retries without exceptions (as you have in the `analyze_response`), we would instead need to use a different form, like:

~~~ ruby
  def send_events(events)
    # build and compress the body just once
    body = LogStash::Json.dump(events.map(&:to_hash))
    compressed_body = gzip(body)

    loop do
      begin
        response = @load_balancer.select do |selected_host_uri|
          # use `Manticore::Client#post!(...)` instead of `Object#send("post",...).call`
          client.post(selected_host_uri, :body => compressed_body, :headers => @headers)
        end
        # break from loop if success
        # break from loop if shutdown requested
        # else allow loop to repeat
      rescue => exception
        # break from loop if shutdown requested
        # else allow loop to repeat
      end
    end
  end
~~~

This form has some repetition, requires that both the non-exception handling and the exception handling know about the requested shutdown state (especially as-implemented with the `analyze_*` methods logging about whether or not a retry will occur), and does not have a way of making non-exceptional retriable 5xx responses deprioritise the selected downstream host from the load balancer.

A form that would _squash_ the response and exception into a single handleable object would look something like:

~~~ ruby
  def send_events(events)
    # build and compress the body just once
    body = LogStash::Json.dump(events.map(&:to_hash))
    compressed_body = gzip(body)

    loop do
      response_or_exception = @load_balancer.select do |selected_host_uri|
        # use `Manticore::Client#post!(...)` instead of `Object#send("post",...).call`
        client.post(selected_host_uri, :body => compressed_body, :headers => @headers)
      end rescue $!
      # break loop if `response_or_exception` is a success response
      # break loop if shutdown requested
      # break loop if `response_or_exception` is a non-retryable exception
      # break loop if `response_or_exception` is a non-retryable response
    end
  end
~~~

In the end, with our simple load balancer only deprioritising on _exceptions_ raised in the block, we may be best off raising exceptions in that block when receiving a non-2xx response, and unifying on a single handler for exceptions raised (which would go back to the first form I recommended).

Reply:
_(sorry to remove queue late, went through test cases without pushing.)_

Non-repeating and handling cases (response, exception and shutdown) independently, brought me to this pseudo:

```
def send_events(events)
    body = LogStash::Json.dump(events.map(&:to_hash))
    compressed_body = gzip(body)

    loop do
      begin
        response = @load_balancer.select do |selected_host_uri|
          http_client.post(selected_host_uri, :body => compressed_body, :headers => @headers).call # need to use call to make sure we mark error in the Router#select->rescue
        end

        # break loop on success response
        # allow retrying unless pipeline shutdown requested and retriable response
      rescue => exception
        # allow retrying unless pipeline shutdown requested and retriable exception
      end
       # handle shutdown request
    end
  end
```

Reply:
One important note here, as commented on HTTP call request piece: 
- we need to actually send the HTTP request, otherwise we cannot mark error in the Router#select->rescue
- without using `.call`, the select block returns a future which _actual_ call will happen when we interact with future (ex: `response.code`)